### PR TITLE
Add type annotations to everything referenced in __init__.py

### PR DESCRIPTION
**Pull request by csmarchbanks**
Fixes: #760 

This should cover most of the public surface area used by others. There is more typing work to be done before we can enable stricter types, but this will be a good next step.

### CODE REVIEW

File: prometheus_client/exposition.py [hunk 1]
```diff
@@ -271,27 +272,34 @@ def factory(cls, registry):
...
         os.replace(tmppath, path)
     else:
         os.rename(tmppath, path)
 
 
-def _make_handler(url, method, timeout, headers, data, base_handler):
+def _make_handler(
+    url: str,
+    method: str,
+    timeout: Optional[float],
```
Comment (tommyjcarpenter) [inline prometheus_client/exposition.py:294]: Seems fine to me, but most cases where I see `Optional`, there is typically a default of `=None`; but in this case I guess the caller of this is explicitly required to pass in an explicit `None`..

File: prometheus_client/exposition.py [hunk 2]
```diff
@@ -343,8 +357,13 @@ def handle():
 
 
 def push_to_gateway(
-        gateway, job, registry, grouping_key=None, timeout=30,
-        handler=default_handler):
+    gateway: str,
+    job: str,
+    registry: CollectorRegistry,
+    grouping_key: Optional[Dict[str, str]] = None,
```
Comment (tommyjcarpenter) [inline prometheus_client/exposition.py:363]: if grouping key is reused below, if it has a constant structure, may want to use a `TypedDict`

File: prometheus_client/exposition.py [hunk 3]
```diff
@@ -343,8 +357,13 @@ def handle():
...
+    timeout: Optional[float] = 30,
```
Comment (tommyjcarpenter) [inline prometheus_client/exposition.py:364]: here timeout has a default.

File: prometheus_client/metrics_core.py [hunk 4]
```diff
@@ -169,10 +201,17 @@ def __init__(self, name, documentation, count_value=None, sum_value=None, labels
         if labels is None:
             labels = []
         self._labelnames = tuple(labels)
```
Comment (tommyjcarpenter) [inline prometheus_client/metrics_core.py:203]: maybe `self._labelnames = tuple(labels) if labels else ()`

File: prometheus_client/samples.py [hunk 5]
```diff
@@ -4,30 +4,30 @@
...
 
-    def __init__(self, sec, nsec):
+    def __init__(self, sec: Union[int, float], nsec: Union[int, float]) -> None:
         if nsec < 0 or nsec >= 1e9:
             raise ValueError(f"Invalid value for nanoseconds in Timestamp: {nsec}")
         if sec < 0:
             nsec = -nsec
-        self.sec = int(sec)
-        self.nsec = int(nsec)
+        self.sec: int = int(sec)
```
Comment (tommyjcarpenter) [inline prometheus_client/samples.py:12]: did mypy complain here when theres an explicit int() cast but it didnt infer the type fro that?

**Review by tommyjcarpenter — APPROVED** [review]
thanks so much!

File: prometheus_client/exposition.py (hunk 2 above)
Comment (csmarchbanks) [reply prometheus_client/exposition.py:363]: I don't think it it has a constant structure, and `TypedDict` isn't available earlier than 3.8 anyway :(. I really want to use `Protocol` in a few places as well but cannot yet.

File: prometheus_client/exposition.py (hunk 1 above)
Comment (csmarchbanks) [reply prometheus_client/exposition.py:294]: Yeah, this one is an internal and requires that we explicitly pass timeout though even if the value is None (or a default value) from one of the other handlers. A bit unusual I agree.

File: prometheus_client/metrics_core.py (hunk 4 above)
Comment (csmarchbanks) [reply prometheus_client/metrics_core.py:203]: Decent idea as a future cleanup, going to leave it out of this PR to keep the focus on types clear :).

File: prometheus_client/samples.py (hunk 5 above)
Comment (csmarchbanks) [reply prometheus_client/samples.py:12]: Ahh, good catch. I think this was from a test with pyright which somehow didn't infer it from the cast. We aren't using pyright yet so I removed them.

**Review by csmarchbanks — COMMENTED** [review]
Thanks for the review!

File: prometheus_client/exposition.py (hunk 2 above)
Comment (tkukushkin) [reply prometheus_client/exposition.py]: It looks like you can use typing-extensions for `TypedDict` and `Protocol`.

File: prometheus_client/samples.py (hunk 5 above)
Comment (tkukushkin) [reply prometheus_client/samples.py]: It is a good idea to describe the types of public fields in the body of the class. Other users can use pyright for type checking their projects and pyright doesn't trying to infer types in such cases (`__init__` can be much more complicated), in py.typed project all public members should be type annotated.

Comment (csmarchbanks) [reply prometheus_client/samples.py:12]: Interesting, thanks for the link! I updated this code back to having explicit int types and looked through the other classes assigning to public variables in this PR and it looks like they are covered. Hopefully that helps out pyright users while we work on getting more coverage.

**Pull request merged**